### ISSUE TITLE
 'When' clauses are ignored for all commands in global menu

### ISSUE THREAD

**Issue by monk-time** [Microsoft/vscode#35225]
- VSCode Version: Code - Insiders 1.17.0-insider (f7962f0682a76516df51de4856f8ccc5d8ad578a, 2017-09-20T05:24:33.599Z)
- OS Version: Windows_NT x64 6.1.7601
- Extensions: none
---

Key bindings for commands that toggle sidebar or panel visibility ignore 'when' clause contexts, making it impossible to limit them to certain contexts.

**Use case:** some shortcut schemes (e.g. Intellij IDEA) have a shortcut to hide current active sidebar or panel, and [a VScode extension](https://marketplace.visualstudio.com/items?itemName=k--kato.intellij-idea-keybindings) that provides such keybindings wants to recreate this functionality [through a shortcut that toggles panels and the sidebar only when they are active](
https://github.com/k--kato/vscode-intellij-idea-keybindings/blob/196930a21d640292fba481200489234a272220e1/package.json#L771-L807), essentially turning a toggle into a closer. But since 'when' clause contexts for these actions always evaluate to true, the shortcuts become globally active and act as toggles. And while the last three shortcuts from that snippet are replaceable with a global shortcut for `workbench.action.closePanel`, there is no corresponding action to close (not toggle) the sidebar.

**Steps to reproduce:**

1. Pick an unused shortcut, e.g. `ctrl+l`, and a context that would be inactive during the next steps, e.g. `parameterHintsVisible`.
2. Add any one of the following settings to `keybindings.json`:
```json
{ "key": "ctrl+l",                "command": "workbench.action.toggleSidebarVisibility",
                                     "when": "parameterHintsVisible" }
```
```json
{ "key": "ctrl+l",                "command": "workbench.action.togglePanel",
                                     "when": "parameterHintsVisible" }
```
```json
{ "key": "ctrl+l",                "command": "workbench.actions.view.problems",
                                     "when": "parameterHintsVisible" }
```
```json
{ "key": "ctrl+l",                "command": "workbench.debug.action.toggleRepl",
                                     "when": "parameterHintsVisible" }
```
```json
{ "key": "ctrl+l",                "command": "workbench.action.terminal.toggleTerminal",
                                     "when": "parameterHintsVisible" }
```
3. Set focus to the current text editor.
4. Press `ctrl+l`.

**Expected result:** nothing should happen, the 'when' clause evaluates to false so the shortcut should be disabled.
**Actual result:** the shortcut works, the relevant panel/sidebar visibility is toggled.

**Comment by usernamehw** [comment]
@sandy081 What is this issue: `bug`? `as-designed`? `feature-request`? Is there any milestone?

**Comment by jrieken** [comment]
@monk-time The `parameterHintsVisible`-clause is set within an editor and its value represents if the signature/parameter help widget is visible or not. That usually happens when typing code, e.g function invocations. So, my question is why you depend on that clause?

**Comment by usernamehw** [comment]
I am not the issue author, but previous title was more precise: it doesn't limited to `parameterHintsVisible` context --- these commands ignore **ALL** when contexts (at least the ones that I tried).

This works:
```json
{
    "key": "ctrl+shift+9",
    "command": "editor.action.insertSnippet",
    "when": "explorerViewletVisible"
}
```
This doesn't:
```json
{
    "key": "ctrl+shift+8",
    "command": "workbench.actions.view.problems",
    "when": "explorerViewletVisible"
}
```

**Comment by monk-time** [comment]
@jrieken The poster above is right, and I admit I could've made it clearer: when I was testing this, *all* clauses that I've tried were ignored. As I wrote in Step 1, `parameterHintsVisible` is just an example of a clause that is most likely to be false, it could be any other clause. 

I'll change the title back if you don't mind.

**Comment by usernamehw** [comment]
Is this issue related https://github.com/Microsoft/vscode/issues/26318?

**Comment by jrieken** [comment]
> when I was testing this, all clauses that I've tried were ignored

You most likely tested with when-clauses that are only set inside an editor. There can be multiple editors at the same time (side by side, nested, diff etc.) and therefore multiple (nested) containers for context keys exist. The `parameterHintsVisible`-clause is set multiple times (once per editor) but not "globally" because they only make sense in the scope of an editor. What should the global-value of a editor-local clause be if its `true` for one and `false` for another editor? The rule of thumb is that clauses starting with `editor` are available inside editors only. However we aren't very good with that rule and clauses like `parameterHintsVisible` and others exist... This list list clauses and the scope in which they are available: https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts

**Comment by monk-time** [comment]
@jrieken, I don't have time to understand this undocumented local/global scope distinction, but this issue is about 'when' clauses that don't work as documented. Since it doesn't seem like you have tried to reproduce what I've described in the OP, I took the fresh Insiders build and put a simple action into `keybindings.json`:

    [
        { "key": "ctrl+l",                "command": "editor.action.toggleMinimap",
                                             "when": "inDebugMode" }
    ]

I create a new file, set the cursor to the main and only editor, *debug mode* is obviously not active, I hit `ctrl+l`, and the minimap gets toggled. This means that the clause was completely ignored.

If you are talking about some other issue, please stop changing this one into something that is not related to it and open another issue. Sorry, but I am getting really frustrated here.

**Comment by usernamehw** [comment]
Can you find even 1 when context for which `workbench.action.terminal.toggleTerminal` won't ignore it? I tried: 
[`findWidgetVisible`, `inQuickOpen`, `explorerViewletVisible`, `terminalFocus`, `searchViewletVisible`, `config.editor.minimap.enabled`, `replaceActive`, `globalMessageVisible`, `parameterHintsVisible`, `renameInputVisible`, `openEditorsFocus`, `explorerViewletFocus`, `inDebugMode`] 
and some negations.

**Comment by jrieken** [comment]
> nothing should happen, the 'when' clause evaluates to false so the shortcut should be disabled.

@monk-time I am sorry, I mis read that and though about the opposite. 

I believe the problem is that those command appear in the global menu (with their assigned shortcut) and that therefore they are bypassing our context key checks... It's actually not us handling the keybinding at that point... Needs some thinking

**Comment by jrieken** [comment]
Yep, command is being called by global menu, not the keybindings service.... @bpasero Is this something new?

<img width="903" alt="screen shot 2017-11-10 at 15 00 44" src="https://user-images.githubusercontent.com/1794099/32661640-f99050d0-c627-11e7-8cd4-b555b1359383.png">

**Comment by bpasero** [comment]
@jrieken I verified that this issue exists back until 1.15, so I do not think this is a regression. However it looks like on all platforms, if you press the keybinding of an existing menu item, we go through the menu (first I thought this would only happen on macOS, but seems to happen on Linux and Windows as well).

Unfortunately I see no way of finding out if a command was triggered by explicit click in the menu vs. using the keybinding. If someone has ideas, let me know.

**Comment by jrieken** [comment]
> Unfortunately I see no way of finding out if a command was triggered by explicit click in the menu vs. using the keybinding. If someone has ideas, let me know.

Well, we dispatch it from the main onto the active window. We just don't treat it has key key event but call executeCommand directly

**Comment by bpasero** [comment]
@jrieken yeah that could work, if you can advise what to do where we receive the command in [`vscode:runAction`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/window.ts#L124) we could wire it in. 

I guess what we need to have is the actual keybinding that was added to the menu though, otherwise we could probably not know which `when` clause to evaluate. We get the keybindings from the renderer in [`resolveKeybindings`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/window.ts#L372)

**Comment by jrieken** [comment]
You could try `lookupKeybinding` but I believe that returns `undefined` for both cases: a command has no keybinding or the context doesn't match...

**Comment by peppy** [comment]
If this is to be left how it is, I would at least recommend a warning/error is displayed when a `"when"` scope is used on an unsupported action.

**Comment by bpasero** [comment]
This should work with @sbatten work once it landed.

**Comment by Tyriar** [comment]
@bpasero is the custom title bar the final fix for this though? I don't see a custom title bar even feeling that great on Linux.

**Comment by bpasero** [comment]
@Tyriar what @sbatten is doing will make it also better for the native menu: all entries will be built via commands and update properly.

**Comment by Tyriar** [comment]
@bpasero so the keybindings will update when contexts change?

**Comment by bpasero** [comment]
@Tyriar no, menu items can show up as disabled or removed if the context does not match anymore. this allows to contribute an entry to the main menu that is only enabled/visible under certain conditions (something that we did not have before).

**Comment by bpasero** [comment]
This is working now.

**Comment by bpasero** [comment]
Blocked by dynamic menus still not being enabled on macOS and Linux.

**Comment by sbatten** [comment]
As discussed this will be unblocked debt week in october

**Comment by sbatten** [comment]
resolved with #59936 and #59855

**Comment by chrmarti** [comment]
Still works:
```
	{
		"key": "ctrl+l",
		"command": "workbench.action.toggleSidebarVisibility",
		"when": "false"
	}
```

VS Code version: Code - Insiders 1.29.0-insider (d7ac6e838f51a03f7d417b4758f5fb075878d615, 2018-11-01T06:11:43.390Z)
OS version: Darwin x64 17.7.0

**Comment by alexandrudima** [comment]
@sbatten As we are all aware from multiple discussions on this topic, contexts are different depending on where the focus sits. I can implement a new method in `IKeybindingService` to look up keybindings based on context, but then won't you have 100% the same problem as the one described in #55453 , which cannot be solved due to performance concerns -- specifically https://github.com/Microsoft/vscode/issues/55453#issuecomment-414757635

**Comment by sbatten** [comment]
@alexandrudima this iteration, I have made some improvements that greatly reduce the amount of updates the menu based on context. My current view is that the global application menu reflects the global context so enablement/disablement/keybindings must be based on context keys in the global scope. Thoughts?

**Comment by alexandrudima** [comment]
@sbatten 

> My current view is that the global application menu reflects the global context so enablement/disablement/keybindings must be based on context keys in the global scope. Thoughts?

That is also incorrect, but in a different, more subtle way, and arguably would make things a lot worse. Before jumping to implement something worse, let's look at the problem. 

Take, for example, the Selection menu:

<img src="https://user-images.githubusercontent.com/5047891/48065079-0054c900-e1ca-11e8-8d81-0f6180c90c59.png" width="200" />

Now let's look at a default keybinding from there:

```json
{ "key": "shift+alt+up",          "command": "editor.action.copyLinesUpAction",
                                     "when": "editorTextFocus && !editorReadonly" },
```

`editorTextFocus` is undefined in the global context, so the `when` clause would not hold. So, the action would get no keybinding rendered next to it.

---

Also, I think this bug affects only the native menu, because in the native menu case the menu itself (Electron via C++ code) listens to those keybindings and executes the menu item directly, outside of our logic. For custom menus, this is no longer a problem.

---

I think there are two possible correct fixes:
1) evaluate focused context
2) enhance Electron to provide information if a native menu item is invoked because the user clicked on it, or used arrow keys + enter, or invoked the keyboard shortcut. If we were to differentiate between these cases, then we could interpret ourselves the keyboard shortcut.

**Comment by jrieken** [comment]
> enhance Electron to provide information if a native menu item is invoked because the user clicked on it, or used arrow keys + enter, or invoked the keyboard shortcut. 

We might be able to know that. I have tested mac and clicking vs keyboard shortcuts can be differentiated by the event - it's only partial but when pressing `cmd+z` the `metaKey` modifier is set, when clicking all are false. 

<img width="1555" alt="screenshot 2018-11-06 at 15 25 14" src="https://user-images.githubusercontent.com/1794099/48070639-a27bad80-e1d8-11e8-8713-3b1172789052.png">

**Comment by sbatten** [comment]
@alexandrudima 
>  My current view is that the global application menu reflects the global context so enablement/disablement/keybindings must be based on context keys in the global scope.

This was more of a statement of how I think things should be not necessarily how they are, but I agree this should be fleshed out. With this view actually in place, it limits the menu updates to relevant updates only for enablement/disablement/keybindings thanks to the support in @jrieken's menu model.

For keybindings, we *could* fix this issue by trying to update the menu on every focus change, but that seems excessive. We also would still run into this issue with enablement/disablement, however. Keybindings get resolved for the menu each time it is updated. Enablement/Disablement does not. The context is set once when the menus are created in the model.

> lookup keybindings based on focused context and constantly update the menus.

This makes me ask the question why the context service doesn't support evaluating against the current context based on focus if that's what matters instead of being required to track focus around. If I could set a flag on the context service to do this before handing it to the menu model and the keybinding service could use that when evaluating keybinds, it could work.

Yes, I am coupling two issues but they are both related to resolving commands against a focused context.
